src/uuid/rfc4122.cr

```diff
@@ -0,0 +1,85 @@
+# Support for RFC 4122 UUID variant.
+struct UUID
+  # RFC 4122 UUID variant versions.
+  enum Version
+    # Unknown version.
+    Unknown = 0
+
+    # Version 1 - date-time and MAC address.
+    V1 = 1
+
+    # Version 2 - DCE security.
+    V2 = 2
+
+    # Version 3 - MD5 hash and namespace.
+    V3 = 3
+
+    # Version 4 - random.
+    V4 = 4
+
+    # Version 5 - SHA1 hash and namespace.
+    V5 = 5
+  end
+
+  # Generates RFC 4122 UUID `variant` with specified `version`.
+  def initialize(version : Version)
+    case version
+    when Version::V4
+      @data = SecureRandom.random_bytes(16)
+      variant = Variant::RFC4122
+      version = Version::V4
+    else
+      raise ArgumentError.new "Creating #{version} not supported."
+    end
+  end
+
+  # Returns version based on provided 6th `byte` (0-indexed).
+  def self.byte_version(byte : UInt8)
+    case byte >> 4
+    when 1 then Version::V1
+    when 2 then Version::V2
+    when 3 then Version::V3
+    when 4 then Version::V4
+    when 5 then Version::V5
+    else        Version::Unknown
+    end
+  end
+
+  # Returns byte with encoded `version` for provided 6th `byte` (0-indexed) for known versions.
+  # For `Version::Unknown` `version` raises `ArgumentError`.
+  def self.byte_version(byte : UInt8, version : Version) : UInt8
```

Comment:
What about `encode_version`? But it's used once, so maybe it should be inlined or made private (and doesn't have to raise)?